src/main/scala/org/broadinstitute/dsde/agora/server/dataaccess/acls/gcs/GcsAuthorizationProvider.scala

```diff
@@ -2,91 +2,180 @@
+    val url = createBucketAclUrl(entity)
+    val content = GoogleAccessControls(kind   = Option(GoogleAccessControls.bucketKind),
+                                       entity = Option(prefixUser(username)),
+                                       role   = Option(GcsRole.Owner))
+    val response = gcsPOST(url, content)
+
+    response.onSuccess { case r: HttpResponse =>
+      if (r.status == OK) bucketAcl.success(r)
+      else throw GoogleConnectionException()
     }
+    response.onFailure { case _ => throw GoogleConnectionException() }
   }
 
-  override def authorizationsForEntity(agoraEntity: Option[AgoraEntity], username: String): AuthorizedAgoraEntity = {
-    agoraEntity match {
-      case None =>  AuthorizedAgoraEntity(None, new AgoraPermissions(Nothing))
-      case Some(entity) =>
-        val responseFuture: Future[AuthorizedAgoraEntity] =
-          pipeline(Get(getGcsObjectRoleRequestUrl(entity, username))).map(response =>
-            AuthorizedAgoraEntity(agoraEntity, RoleTranslator.gcsObjectToEntityPermissions(unrollGaclResponse(response))))
-        Await.result(responseFuture, timeout)
+  def createObject(entity: AgoraEntity, username: String, _object: Promise[HttpResponse]): Unit = {
+    val url = createObjectUrl(entity)
+    val response = gcsPOST(url)
+
+    response.onSuccess { case r: HttpResponse =>
+      if (r.status == OK) _object.success(r)
+      else throw GoogleConnectionException()
     }
+    response.onFailure { case _ => throw GoogleConnectionException() }
   }
 
-  override def authorizationsForEntities(agoraEntities: Seq[AgoraEntity], username: String): Seq[AuthorizedAgoraEntity] = {
-    // Get a sequence of response Futures, one for each Some(entity). Transform the entity + HttpResponse -> AuthorizedAgoraEntity
-    // Transform the sequence of Futures into a Future of a sequence
-    // Futures can complete in parallel. We transform them as they complete. Overall, we await the result on the slowest HttpResponse
-    val responseFutures: Future[Seq[AuthorizedAgoraEntity]] = Future.sequence(agoraEntities.map(entity => {
-      pipeline(Get(getGcsObjectRoleRequestUrl(entity, username))).map(httpResponse =>
-        AuthorizedAgoraEntity(Option(entity), RoleTranslator.gcsObjectToEntityPermissions(unrollGaclResponse(httpResponse))))
-      }))
-    Await.result(responseFutures, timeout)
+  def addUserAclToObject(entity: AgoraEntity, username: String): Unit = {
+    val url = createObjectAclUrl(entity)
+    val content = GoogleAccessControls(kind   = Option(GoogleAccessControls.objectKind),
+                                       entity = Option(prefixUser(username)),
+                                       role   = Option(GcsRole.Owner))
+    val response = gcsPOST(url, content)
+
+    response.onSuccess { case r: HttpResponse => if (r.status != OK) throw GoogleConnectionException() }
+    response.onFailure { case t: Throwable => throw GoogleConnectionException() }
+  }
```

Comment:
We are calling post on url and content and handling success/failure in three separate places. Seems like we could consolidate this logic.